### PR TITLE
initial version of mock service.

### PR DESCRIPTION
**Pull request by jn9e9**
Basic version, loading test files from folder and returning responses if the request matches.

Signed-off-by: john 9e9 <jn9e9e9@gmail.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -157,7 +169,15 @@ The whole test spec is defined in the spec: object in the file.  In that spec, t
 | response.auth.type | This can either be ```none```, which will cause the auth section of the message to be empty (equivalent of the No Authentication type of authentication); or it can be ```direct```, which will cause the auth section of the message to contain authentication data corresponding to the format for Direct Authentication.  If ```direct``` is set, then the result.auth.app_name field must be set.  Note that this field does not cause the header auth_type field to be set.  **NOTE:**  A Parsec client would not send authentication data in a result message, but this spec format allows test authors to create the message as they wish to excersice the parsec client code. |
 | response.auth.app_name | Used to populate the auth section of the message when ```direct``` authentication is selected |
 
+# Mock service testing
```
Comment (hug-dev) [inline README.md:172]: Testing the tester 🤯

File: parsec-mock.py [hunk 2]
```diff
@@ -1,2 +1,94 @@
...
+    """Read test specs from a folder"""
+    specfiles = [f for f in listdir(test_folder) if isfile(join(test_folder, f))]
+
+    for file in specfiles:
+        print(f"Parsing spec file: {file}")
+        with open(os.path.join(test_folder, file), 'r') as f:
+            spec = safe_load(f)
+            testspec = TestSpec(spec)
+            if testspec.is_valid():
+                tests[testspec.test_data.request] = testspec
```
Comment (hug-dev) [inline parsec-mock.py:86]: Why don't we just store `testspec.test_data.response` as values in the dict?

File: parsec-mock.py [hunk 3]
```diff
@@ -1,2 +1,94 @@
...
+
+    while True:
+        connection, client_addr = sock.accept()
+        try:
+            print('Connection received from {}'.format(client_addr))
+            received_data = connection.recv(4096)
+            b64_received_data = base64.b64encode(received_data).decode('ascii')
+            if b64_received_data in test_cases:
+                test_case = test_cases[b64_received_data]
+                print('Received expected request for test case {}'.format(test_case.spec.name))
```
Comment (hug-dev) [inline parsec-mock.py:47]: *referring to my other comment* Ah I see, this is to be able to access to the name of the test.

**Review by hug-dev — APPROVED** [review]
Looks good to me, thanks a lot!

**Comment by hug-dev** [comment]
Ah forgot one thing, do you want to add the `build.yml` workflow to be executed for each PR and add the test you created in this PR there?

File: parsec-mock.py [hunk 4]
```diff
@@ -1,2 +1,94 @@
...
+import base64
+
+from yaml import safe_load
+
+
+@click.command()
+@click.option('--test-folder', default='./testdata', help='Load all tests from folder')
+@click.option('--parsec-socket', default='./parsec.sock', help='Path to parsec unix socket')
+def run_test(test_folder, parsec_socket):
+    print('Mock parsec service listening on unix://{}.'.format(parsec_socket))
```
Comment (ionut-arm) [inline parsec-mock.py:19]: Shouldn't it be `unix:{}`? Though I have to admit, at this point I'm not really sure what the correct format for that scheme is

**Review by ionut-arm — APPROVED** [review]
Looks alright! One nit comment

**Pull request merged**